Make WebKit <input type=checkbox switch> tests iOS-compatible

annevk opened the pull request:
#### 90f2bafbe99f4ab1ec05cd05d975c0bbd9aa3ad4
<pre>
Make WebKit &lt;input type=checkbox switch&gt; tests iOS-compatible
<a href="https://bugs.webkit.org/show_bug.cgi?id=267546">https://bugs.webkit.org/show_bug.cgi?id=267546</a>

Reviewed by Aditya Keerthi.

As there&apos;s no public painting code these shall not pass.

The timeout is increased for a number of tests as the tests will
otherwise fail on iOS builds with painting code. That is because iOS
has a separate animation for the switch&apos;s thumb that depends on the
pressed state and not the pointer position.

* LayoutTests/fast/forms/switch/click-animation-disabled.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted-document-removed.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted-type-change.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted.html:
* LayoutTests/fast/forms/switch/click-animation-preventdefault.html:
* LayoutTests/fast/forms/switch/click-animation-redundant-checked.html:
* LayoutTests/fast/forms/switch/click-animation-redundant-disabled.html:
* LayoutTests/fast/forms/switch/click-animation-twice-fast.html:
* LayoutTests/fast/forms/switch/click-animation-twice.html:
* LayoutTests/fast/forms/switch/click-animation.html:
* LayoutTests/fast/forms/switch/pointer-tracking-disabled.html:
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl.html:
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again.html:
* LayoutTests/fast/forms/switch/pointer-tracking.html:
* LayoutTests/platform/ios/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/273260@main">https://commits.webkit.org/273260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe2c251a6e9bceac59e0b8bd7dbd19e07a69cc77

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34758 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13579 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36766 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37444 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31367 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10664 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30314 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30950 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10047 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10108 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38709 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31553 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31349 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36159 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10226 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8128 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34133 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12054 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30668 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10773 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4488 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11117 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -27,6 +27,22 @@ fast/events/ios [ Pass ]
+fast/forms/switch/click-animation-twice.html [ ImageOnlyFailure ]
+fast/forms/switch/click-animation.html [ ImageOnlyFailure ]
+fast/forms/switch/pointer-tracking-there-and-back-again-rtl.html [ Skip ]
+fast/forms/switch/pointer-tracking-there-and-back-again.html [ Skip ]
```
Is it intentional that these two tests aren't also being updated to use `UIHelper` in this patch?

annevk replied:
Yeah, UIHelper doesn't have helpers for these as far as I can tell.

pxlcoder replied:
I think you can rewrite those tests to use `UIHelper.EventStreamBuilder`.

Example: https://github.com/WebKit/WebKit/blob/main/LayoutTests/fast/events/touch/ios/scroll-on-touch-start-with-slow-event-listener.html

annevk replied:
I have written this code, but I can only get the tests to sometimes pass on a build with painting code. It's rather non-deterministic. Would you mind having a look and see if there's anything obviously wrong? I updated this PR.

pxlcoder replied:
You need to use `async`/`await` whenever using `UIHelper` with user interaction.

annevk replied:
Why? I have the click event to tell me when the interaction ended, no?

pxlcoder commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -27,6 +27,22 @@ fast/events/ios [ Pass ]
 fast/forms/ios/ipad [ Skip ]
+fast/forms/switch [ Pass ]
+
+# More work required for these tests to run on iOS
```
Are all of these due to a lack of painting code, or is there another issue?

annevk replied:
`fast/forms/switch/click-animation-twice.html` only passes if I increase the final timeout from 50 to 75.

annevk committed:
Make WebKit <input type=checkbox switch> tests iOS-compatible
https://bugs.webkit.org/show_bug.cgi?id=267546

Reviewed by Aditya Keerthi.

As there's no public painting code these shall not pass.

The timeout is increased for a number of tests as the tests will
otherwise fail on iOS builds with painting code. That is because iOS
has a separate animation for the switch's thumb that depends on the
pressed state and not the pointer position.

* LayoutTests/fast/forms/switch/click-animation-disabled.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted-document-removed.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted-type-change.html:
* LayoutTests/fast/forms/switch/click-animation-interrupted.html:
* LayoutTests/fast/forms/switch/click-animation-preventdefault.html:
* LayoutTests/fast/forms/switch/click-animation-redundant-checked.html:
* LayoutTests/fast/forms/switch/click-animation-redundant-disabled.html:
* LayoutTests/fast/forms/switch/click-animation-twice-fast.html:
* LayoutTests/fast/forms/switch/click-animation-twice.html:
* LayoutTests/fast/forms/switch/click-animation.html:
* LayoutTests/fast/forms/switch/pointer-tracking-disabled.html:
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl.html:
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again.html:
* LayoutTests/fast/forms/switch/pointer-tracking.html:
* LayoutTests/platform/ios/TestExpectations:

Canonical link: https://commits.webkit.org/273260@main
